### DISCUSSION TITLE
How to run two instances of VWsFriend on the same machine?

### DISCUSSION THREAD

**Discussion by oldgitdaddy** [tillsteinbach/VWsFriend#576] [Q&A]
I want to run two separate instances of VWsfriend, querying two different VW accounts. I have already changed some of the ports, but I still get some weird interferences. For instance the server to query the VW API is not properly available.

DB ports in my second setup:

```
VWSFRIEND_PORT=6000
GF_SERVER_HTTP_PORT=5000
DB_PORT=5431
```
What did I miss? Or am I totally on the wrong track and it is much easier?

---
UPDATE:
So I did some further analysis and got the service up and running, sort of.
I figured that vwsfriend server is running fine, just the port mapping is not working properly. With this it at least collects the data from weconnect. I completed the onboarding data by adjusting the postgresql port mapping in the docker-compose and inserting the settings for the app and the vehicle. Still the server on port 6000 is not available and my original question is still valid, but probably not this urgent, as I have a workaround in place. 

I think it would be good to fix this issue and also to add some guide for development setup. Thank you!

**Comment by tillsteinbach** [comment]
No, I think basically you are doing it right. Any error messages?

**Comment by oldgitdaddy** [comment]
There are no errors as far as I can tell. `docker-compose logs` just returns usual stuff. Anyway it is strange, that the additional 3 services do not show up via `docker-compose ps`.

Probably I mixed things up. Is there a way to be sure that the database has been properly deleted? `docker-compose down` should delete the instances and the physical data base, right?

**Comment by oldgitdaddy** [comment]
To be precise situation is as follows:
Instance 1 is working fine
Instance 2 has the following behavior:
- Weirdly login to grafana dashboard of instance 2 with user from instance 1 is working, the user for instance 2 is not working
- Grafana of instance 2 shows vehilces from instance 1 and instance 2
- VWFriend server calling VW API is not available through browser.

**Reply by tillsteinbach** [reply]
I would try to have a clean start with the instance two. `docker-compose down` does not delete the volume. `docker-compose down -v` should

**Reply by oldgitdaddy** [reply]
You are right. Thanks. This partially solved my problem. Still I cannot access the settings for the vwfriend server. Further I get this warning in grafana dashboard.

![image](https://github.com/tillsteinbach/VWsFriend/assets/4282985/d263b61d-9f61-4fad-b860-d351b992db97)

**Reply by oldgitdaddy** [reply]
Not sure if the use of EXPOSE needs to be parameterized as well... Currently it is exposing hardcoded port 4000.

[https://github.com/tillsteinbach/VWsFriend/blob/main/vwsfriend/Dockerfile#L56](https://github.com/tillsteinbach/VWsFriend/blob/main/vwsfriend/Dockerfile#L56)

**Reply by tillsteinbach** [reply]
The warning will be gone when you have set the settings. You can change those by going here: `http://<IPOFSECONDINSTANCE>: <VWSFRIEND_PORT>/database/settings/edit`

**Reply by oldgitdaddy** [reply]
Yes. But I cannot access the VWFriend server on port 6000 in my case due to network timeout. Also not sure how to access any designated logs of the service, How do I check, if the server is up anyway?

**Reply by oldgitdaddy** [reply]
Ok. I see now that docker-compose is not always listing all services. You have to name the explicit service. 
This is what docker-compose ps returns:
Instance 1:  `Up (healthy)   0.0.0.0:4000->4000/tcp,:::4000->4000/tcp`
Instance 2:  `Up (healthy)   4000/tcp, 0.0.0.0:6000->6000/tcp,:::6000->6000/tcp`

Any ideas, Why instance 2 is not available then on port 6000?

**Comment by tillsteinbach** [comment]
> Not sure if the use of EXPOSE needs to be parameterized as well... Currently it is exposing hardcoded port 4000.
> 
> https://github.com/tillsteinbach/VWsFriend/blob/main/vwsfriend/Dockerfile#L56

hmmm, not sure, so far I thought the EXPOSE in the Dockerfile is more a documentation thingy

**Reply by oldgitdaddy** [reply]
Or it may be the Dockerfile here: https://github.com/tillsteinbach/VWsFriend/blob/d8c06718c24f45853bedc73cbf791fadef3a2815/vwsfriend/Dockerfile#L40
What do you think?

**Comment by oldgitdaddy** [comment]
I updated the original description, as I found a workaround by connecting to db and inserting the onboarding data manually.

BTW: This is still the message in grafana
![image](https://github.com/tillsteinbach/VWsFriend/assets/4282985/ad1f7dcb-2a71-4f20-95b3-56f1f1ea8eee)
